src/job.rs

```diff
@@ -1,56 +1,99 @@
 use latch::Latch;
 use std::mem;
+#[cfg(feature = "nightly")]
+use std::any::Any;
+#[cfg(feature = "nightly")]
+use std::panic::{self, AssertRecoverSafe};
+
+enum JobResult<T> {
+    None,
+    Ok(T),
+    #[cfg(feature = "nightly")]
+    Panic(Box<Any + Send>),
+}
 
 /// A `Job` is used to advertise work for other threads that they may
 /// want to steal. In accordance with time honored tradition, jobs are
 /// arranged in a deque, so that thieves can take from the top of the
 /// deque while the main worker manages the bottom of the deque. This
 /// deque is managed by the `thread_pool` module.
-pub struct Job<L:Latch> {
-    /// code to execute (if job is stolen)
-    code: *mut Code,
+pub trait Job<L:Latch> {
+    unsafe fn execute(&mut self);
+}
 
-    /// latch to signal once execution is done (if job is stolen)
-    latch: *mut L,
+pub struct JobImpl<L:Latch,F,R> {
+    pub latch: L,
+    func: Option<F>,
+    result: JobResult<R>,
 }
 
-impl<L:Latch> Job<L> {
-    pub unsafe fn new<'a>(code: *mut (Code+'a), latch: *mut L) -> Job<L> {
-        let code: *mut Code = mem::transmute(code);
-        Job {
-            code: code,
-            latch: latch
+impl<L:Latch,F,R> JobImpl<L,F,R>
+    where F: FnOnce() -> R
+{
+    pub fn new(func: F, latch: L) -> JobImpl<L,F,R> {
+        JobImpl {
+            latch: latch,
+            func: Some(func),
+            result: JobResult::None,
         }
     }
 
-    pub unsafe fn execute(&mut self) {
-        (*self.code).execute();
-        (*self.latch).set();
+    pub unsafe fn as_static(&mut self) -> *mut Job<L> {
+        mem::transmute(self as *mut Job<L>)
     }
-}
 
-pub trait Code {
-    unsafe fn execute(&mut self);
-}
+    pub fn run_inline(self) -> R {
+        self.func.unwrap()()
+    }
 
-pub struct CodeImpl<F,R> {
-    func: Option<F>,
-    dest: *mut Option<R>,
-}
+    #[cfg(not(feature = "nightly"))]
+    pub fn into_result(self) -> R {
+        match self.result {
+            JobResult::None => panic!("job function panicked"),
+            JobResult::Ok(x) => x,
+        }
+    }
 
-impl<F,R> CodeImpl<F,R>
-    where F: FnOnce() -> R
-{
-    pub fn new(func: F, dest: *mut Option<R>) -> CodeImpl<F,R> {
-        CodeImpl { func: Some(func), dest: dest }
+    #[cfg(feature = "nightly")]
+    pub fn into_result(self) -> R {
+        match self.result {
+            JobResult::None => unreachable!(),
+            JobResult::Ok(x) => x,
+            JobResult::Panic(x) => panic::propagate(x),
+        }
+    }
+
+    #[cfg(not(feature = "nightly"))]
+    fn run_result(func: F) -> JobResult<R> {
+        JobResult::Ok(func())
+    }
+
+    #[cfg(feature = "nightly")]
+    fn run_result(func: F) -> JobResult<R> {
+        // RecoverSafe is a bit ugly to deal with...
+        // Ideally this would just be panic::recover(func)
+        let mut wrapper = AssertRecoverSafe::new(Some(func));
```

Comment:
It'd be good to add a comment explaining why we think it's legit to call this function "recover safe" despite having no idea what kind of data it touches. (The answer of course is because we propagate panics.) So perhaps just write: "// We can call this recover safe because we propagate panics."